docs/docs/configuration/settings.md
---
sidebar_position: 2
---

# 🔧 Settings

OctoType accepts a configuration file in the [TOML](https://toml.io/en/) format,
located in the default configuration folder for your system:

| System  | Location                                                                        |
| ------- | ------------------------------------------------------------------------------- |
| Linux   | `$XDG_CONFIG_HOME/octotype/config.toml` or `$HOME/.config/octotype/config.toml` |
| MacOS   | `$HOME/Library/Application Support/com.Mahlquist.OctoType/config.toml`          |
| Windows | `%AppData%\OctoType\config\config.toml`                                         |

The configuration location can also be specified with the `-c <path>` flag.

Some official themes can be found
[in the repo](https://github.com/mahlquistj/octotype/tree/main/themes)

## Default settings

The default settings looks like so:

```toml
sources_dir = "<OCTOTYPE_CONFIG_DIR>/sources"
modes_dir = "<OCTOTYPE_CONFIG_DIR>/modes"
words_per_line = 5
show_ghost_lines = 3
ghost_opacity = [
    0.2,
    0.5,
    0.8,
]
disable_ghost_fade = false

[theme]
term_fg = "<YOUR_TERMINAL_FOREGROUND_COLOR>"
term_bg = "<YOUR_TERMINAL_BACKGROUND_COLOR>"

[theme.spinner]
color = "Yellow"
animation = [
    "⣷",
    "⣯",
    "⣟",
    "⡿",
    "⢿",
    "⣻",
    "⣽",
    "⣾",
]
timing_millis = 250

[theme.text]
success = "#A6E3A1"
warning = "#F9E2AF"
error = "#F38BA8"
highlight = "#89B4FA"

[theme.plot]
raw_wpm = "#CDD6F4"
actual_wpm = "#F9E2AF"
accuracy = "#CDD6F4"
errors = "#F38BA8"
scatter_symbol = "Dot"
line_symbol = "HalfBlock"

[theme.cursor]
color = "White"
text = "Black"

[statistic]
save_enabled = true
history_limit = 10
```

## Options and Types

| Key                           | Type         | Description                                                                                             |
| ----------------------------- | ------------ | ------------------------------------------------------------------------------------------------------- |
| `sources_dir`                 | `String`     | Overwrites the path of the sources directory                                                            |
| `modes_dir`                   | `String`     | Overwrites the path of the modes directory                                                              |
| `words_per_line`              | `int`        | How many words should be displayed per line                                                             |
| `show_ghost_lines`            | `int`        | How many "ghost lines" should be displayed around the active line                                       |
| `ghost_opacity`               | `[float]`    | Overwrite the levels of opacity for each ghost line. Must have a length matching `show_ghost_lines`     |
| `disable_ghost_fade`          | `bool`       | Set this to true if you want the "scrolling" behaviour of ghost lines, but don't like the fading colors |
| `theme.term_fg`               | `Color`      | The foreground of your terminal (Queried directly from you terminal or else it defaults to White)       |
| `theme.term_bg`               | `Color`      | The background of your terminal (Queried directly from you terminal or else it defaults to Black)       |
| `theme.spinner.color`         | `Color`      | Sets the color of the loading-screen spinner                                                            |
| `theme.spinner.animation`     | `[String]`   | Sets the animation of the loading-screen spinner                                                        |
| `theme.spinner.timing_millis` | `int`        | How many milliseconds per "frame" of the animation                                                      |
| `theme.text.success`          | `Color`      | Sets the color of `success`-type text                                                                   |
| `theme.text.warning`          | `Color`      | Sets the color of `warning`-type text                                                                   |
| `theme.text.error`            | `Color`      | Sets the color of `error`-type text                                                                     |
| `theme.text.highlight`        | `Color`      | Sets the color of highlighted text                                                                      |
| `theme.plot.raw_wpm`          | `Color`      | Sets the color of the raw_wpm datapoints                                                                |
| `theme.plot.actual_wpm`       | `Color`      | Sets the color of the actual_wpm datapoints                                                             |
| `theme.plot.accurracy`        | `Color`      | Sets the color of the accurracy datapoints                                                              |
| `theme.plot.errors`           | `Color`      | Sets the color of the error datapoints                                                                  |
| `theme.plot.scatter_symbol`   | `PlotSymbol` | Sets the symbols of scatter-type plots (errors)                                                         |
| `theme.plot.line_symbol`      | `PlotSymbol` | Sets the symbols of line-type plots (wpm, accurracy, etc.)                                              |
| `theme.cursor.color`          | `Color`      | The color of the cursor when in a session                                                               |
| `theme.cursor.text`           | `Color`      | The color of the text under the cursor                                                                  |
| `statistics.save_enabled`     | `bool`       | (Reserved for future use): Wether statistics should save to disk                                        |
| `statistics.history_limit`    | `int`        | (Reserved for future use): How many session-statistics can be saved before deleting. 0 = infinite       |

### Colors

:::warning

While ANSI colors are supported, OctoType is made for mordern terminals with
24bit color support - Using hex colors is recommended for the best experience
possible.

Some known issues regarding ghost-lines can happen, if not using hex colors.

ANSI-Colors may or may not be deprecated in future releases.

:::

Two different types of colors can be supplied: ANSI or Custom which sets a
specific color.

| Color          | Type   | Description |
| -------------- | ------ | ----------- |
| `#RRGGBB`      | Custom | Hex color   |
| `<integer>`    | ANSI   | ANSI index  |
| `Black`        | ANSI   |             |
| `Red`          | ANSI   |             |
| `Green`        | ANSI   |             |
| `Yellow`       | ANSI   |             |
| `Blue`         | ANSI   |             |
| `Magenta`      | ANSI   |             |
| `Cyan`         | ANSI   |             |
| `Gray`         | ANSI   |             |
| `DarkGray`     | ANSI   |             |
| `LightRed`     | ANSI   |             |
| `LightGreen`   | ANSI   |             |
| `LightYellow`  | ANSI   |             |
| `LightBlue`    | ANSI   |             |
| `LightMagenta` | ANSI   |             |
| `LightCyan`    | ANSI   |             |
| `White`        | ANSI   |             |

### PlotSymbol

This type determines what symbols are used to display data within the graphs

| Name        | Symbol          |
| ----------- | --------------- |
| `Dot`       | `•`             |
| `Block`     | `█`             |
| `HalfBlock` | `█, ▄, and ▀`   |
| `Braille`   | `⠓, ⣇, ⣿, etc.` |
| `Bar`       | `▄`             |


docs/docs/configuration/parameters.md
---
sidebar_position: 3
---

# ⚙️ Parameters

[Sources](sources) and [Modes](modes) both have a `parameters` section where you
can define arbritrary keys.

These parameters can be edited by the user upon selection and can be used to
customize the command that is executed by the [Source](sources)

## Types

### Range

A range of numbers.

```toml
"<param>" = {
    min = 10,     # Optional - Defaults to 0
    max = 100,    # Optional - Defaults to 9223372036854775807 (essentially unbounded)
    step = 10,    # Optional - Defaults to 1
    default = 50, # Optional - Defaults to the value of `min`
}
```

Even though every key is optional, at least `min` or `max` must be defined.

### Selection

A list of selectable value.

```toml
"<param>" = {
    options = ["A", "B", "C"], # Required (Can't be empty),
    default = "C",             # Optional - Defaults to the first value in `options`
}
```

### Boolean

A boolean.

```toml
"<param>" = true
```

### Int

An integer - This cannot be changed by the user (Use [Range](#range) instead, if
needed)

```toml
"<param>" = 42
```

### String

A string - This cannot be changed by the user (Use [Selection](#selection)
instead, if needed)

```toml
"<param>" = "some_string"
```

## Replacements

Parameters exist for one reason only: To make an interface for the user to
customize the default values of a [Source](sources).

Every parameter can be used as a replacement in certain places of a
[Source](sources#parameters) or [Mode](modes#parameters).

A Replacement is defined by a `String` which looks like so:
`"{<parameter_key>}"` - This will translate into the final value of the
`parameter_key`

An example of how this is used would look like so:

```toml
# File: <OCTOTYPE_CONFIG_DIR>/sources/programming.toml
[meta]
name = "Programming"
description = "Generates common code-patterns from some languages"
command = [
    "./programming-generator.py",
    "--lang", "{language}", 
    "--mode", "{mode}", 
    "--amount", "{amount}",
    "--function-lines", "{function_lines}"
]
required_tools = ["python3"]

[parameters]
language = {
    options = ["Python", "C", "Rust"]
}
mode = {
    options = ["Words", "Functions"]
}
amount = {
    min: 1
}
# Set the default function-lines for the "Functions" mode to be 3
# This value is never shown to the user, but can be replaced by a Mode
function_lines = 3
```

In our mode we can do the following:

```toml
# File: <OCTOTYPE_CONFIG_DIR>/modes/rust-function-sprint.toml
[meta]
name = "Function sprint"
description = "Complete functions within a time limit"
allowed_sources = ["Programming"]

[parameters]
# A range for how many functions the user wants in total
functions_amount = {
    min = 1,
    default = 3
}
# A range for how many lines we want in each function
function_lines = {
    min = 3,
    default = 5
}
# Another range for how much time they have to type all the words
time_limit = {
    min = 10
    step = 5
    default = 30
}

# Override some parameters in the "Programming" source parameters
[overrides."Programming"]
mode = "Functions"
amount = "{functions_amount}" # Replace with this modes parameter
function_lines = "{function_lines}" # Replace with this modes parameter
```

Now, if the the `Rust function sprint` mode is chosen, it would present the
following options to be changed before starting the session:

```
functions_amount: 3
function_lines: 5
time_limit: 10
language: "Python"
```


octotype/src/config.rs
use std::{collections::HashMap, path::PathBuf, sync::Arc};

use derive_more::{Deref, From};
use directories::ProjectDirs;
use figment::{
    Figment,
    providers::{Format, Serialized, Toml},
};
use serde::{Deserialize, Serialize};
use thiserror::Error;

pub use mode::ModeConfig;
pub use source::SourceConfig;

use crate::config::{stats::StatisticsConfig, theme::Theme};

pub mod mode;
pub mod parameters;
pub mod source;
pub mod stats;
pub mod theme;

#[derive(Debug, From, Error)]
pub enum ConfigError {
    #[error(
        "Failed to get configuration directory. Please specify the location using the `--config <path>` flag"
    )]
    NoDirectory,

    #[error("Failed to create config directory: {0}")]
    CreateDirectory(std::io::Error),

    #[error("Failed to parse config: {0}")]
    Parse(Box<figment::Error>),

    #[error("Failed to parse sources: {0}")]
    ParseSources(source::SourceError),

    #[error("Failed to parse modes: {0}")]
    ParseModes(mode::ModeError),
}

#[derive(Debug, Deserialize, Serialize)]
pub struct Settings {
    pub theme: theme::Theme,
    pub statistic: stats::StatisticsConfig,
    sources_dir: Option<PathBuf>,
    modes_dir: Option<PathBuf>,
    pub words_per_line: usize,
    pub show_ghost_lines: usize,
    #[serde(default)]
    pub ghost_opacity: Vec<f32>,
    pub disable_ghost_fade: bool,
}

impl Default for Settings {
    fn default() -> Self {
        Self {
            theme: Theme::default(),
            statistic: StatisticsConfig::default(),
            sources_dir: None,
            modes_dir: None,
            words_per_line: 5,
            show_ghost_lines: 3,
            ghost_opacity: get_evenly_spread_values(3),
            disable_ghost_fade: false,
        }
    }
}

#[derive(Clone, Debug, Deref, Default, Serialize)]
pub struct Config(Arc<InnerConfig>);

#[derive(Debug, Default, Serialize)]
pub struct InnerConfig {
    pub settings: Settings,
    pub modes: HashMap<String, ModeConfig>,
    pub sources: HashMap<String, SourceConfig>,
}

impl Config {
    pub fn list_modes(&self) -> Vec<String> {
        let mut modes: Vec<_> = self.modes.keys().map(|key| key.to_string()).collect();

        modes.sort();
        modes
    }

    pub fn list_sources(&self) -> Vec<String> {
        let mut sources: Vec<_> = self.sources.keys().map(|key| key.to_string()).collect();

        sources.sort();
        sources
    }

    pub fn sources_dir(&self) -> &PathBuf {
        self.settings.sources_dir.as_ref().unwrap()
    }

    pub fn get(override_path: Option<PathBuf>) -> Result<Self, ConfigError> {
        // Grab default configuration
        let mut settings = Figment::from(Serialized::defaults(Settings::default()));

        // Check for toml file location
        let config_dir = override_path
            .or_else(|| {
                ProjectDirs::from("com", "OctoType", "OctoType")
                    .map(|dirs| dirs.config_dir().to_path_buf())
            })
            .ok_or(ConfigError::NoDirectory)?;

        // Ensure path exists
        if !config_dir.exists() {
            std::fs::create_dir_all(&config_dir)?;
        }

        let mut settings_toml = config_dir.clone();
        if !settings_toml.ends_with("config.toml") {
            settings_toml.push("config.toml");
        }

        if settings_toml.exists() {
            settings = settings.merge(Toml::file(settings_toml));
        }

        let mut settings: Settings = settings.extract().map_err(Box::new)?;

        let sources_dir = settings.sources_dir.clone().unwrap_or_else(|| {
            let mut dir = config_dir.clone();
            dir.push("sources");
            dir
        });
        let sources = source::get_sources(&sources_dir)?;
        settings.sources_dir = Some(sources_dir);

        let modes_dir = settings.modes_dir.clone().unwrap_or_else(|| {
            let mut dir = config_dir.clone();
            dir.push("modes");
            dir
        });
        let modes = mode::get_modes(&modes_dir)?;
        settings.modes_dir = Some(modes_dir);

        if settings.ghost_opacity.len() != settings.show_ghost_lines {
            settings.ghost_opacity = get_evenly_spread_values(settings.show_ghost_lines);
        }

        Ok(Self(Arc::new(InnerConfig {
            settings,
            sources,
            modes,
        })))
    }
}

fn get_evenly_spread_values(num_items: usize) -> Vec<f32> {
    if num_items == 0 {
        return Vec::new();
    }

    if num_items == 1 {
        return vec![0.5];
    }

    let mut values = Vec::with_capacity(num_items);
    let min_val = 0.2;
    let max_val = 0.8;
    let range: f32 = max_val - min_val;

    for i in 0..num_items {
        let normalized_index = i as f32 / (num_items - 1) as f32;
        let value = range.mul_add(normalized_index, min_val);
        values.push(value);
    }

    values
}
